From 6d2f3bbc4336affe8e5bf7940b2f9cb8cf9f04f9 Mon Sep 17 00:00:00 2001
From: Li Wei <liwei825@huawei.com>
Date: Wed, 15 Feb 2023 16:52:11 +0800
Subject: [PATCH 325/921] [Huawei] ocfs2: add customized modifications for
 resize

Offering: HULK
ocfs2 inclusion
category: feature
bugzilla: NA

----------------------------------------

Add customized modifications for resize.

Signed-off-by: Li Wei <liwei825@huawei.com>
---
 fs/ocfs2/Makefile      |  58 +++++
 fs/ocfs2/ocfs2_ioctl.h |  76 ++++++-
 fs/ocfs2/ocfs2_trace.h |  15 +-
 fs/ocfs2/resize.c      | 477 ++++++++++++++++++++++++++++++++++++-----
 fs/ocfs2/resize.h      |  11 +-
 5 files changed, 573 insertions(+), 64 deletions(-)

diff --git a/fs/ocfs2/Makefile b/fs/ocfs2/Makefile
index e69de29bb2d1..2a34ab3edbb4 100644
--- a/fs/ocfs2/Makefile
+++ b/fs/ocfs2/Makefile
@@ -0,0 +1,58 @@
+# SPDX-License-Identifier: GPL-2.0
+ccflags-y := -Ifs/ocfs2 -DCONFIG_OCFS2_DEBUG_MASKLOG
+
+ifneq ($(CONFIG_X86_32)_$(CONFIG_X86_64), )
+	ccflags-y += -DARCH_TYPE_IS_X86
+endif
+
+ifneq ($(CONFIG_ARM), )
+	ccflags-y += -DARCH_TYPE_IS_ARM
+endif
+
+obj-$(CONFIG_OCFS2_FS) += 	\
+	ocfs2.o			\
+	ocfs2_stackglue.o
+
+ocfs2-objs := \
+	alloc.o 		\
+	aops.o 			\
+	blockcheck.o		\
+	buffer_head_io.o	\
+	dcache.o 		\
+	dir.o 			\
+	adlglue.o		\
+	export.o 		\
+	extent_map.o 		\
+	file.o 			\
+	heartbeat.o 		\
+	inode.o 		\
+	ioctl.o 		\
+	journal.o 		\
+	localalloc.o 		\
+	locks.o			\
+	mmap.o 			\
+	namei.o 		\
+	refcounttree.o		\
+	reservations.o		\
+	move_extents.o		\
+	resize.o		\
+	slot_map.o 		\
+	suballoc.o 		\
+	super.o 		\
+	symlink.o 		\
+	sysfile.o 		\
+	uptodate.o		\
+	quota_local.o		\
+	quota_global.o		\
+	xattr.o			\
+	acl.o	\
+	filecheck.o		\
+	lockglue.o
+
+ocfs2_stackglue-objs := stackglue.o
+
+obj-$(CONFIG_OCFS2_FS) += lockfs/
+# cluster/ is always needed when OCFS2_FS for masklog support
+obj-$(CONFIG_OCFS2_FS) += cluster/
+obj-$(CONFIG_OCFS2_FS_O2CB) += adl/
+CFLAGS_stackglue.o := -fstack-protector-strong
diff --git a/fs/ocfs2/ocfs2_ioctl.h b/fs/ocfs2/ocfs2_ioctl.h
index d7b31734f6be..b23762287bb4 100644
--- a/fs/ocfs2/ocfs2_ioctl.h
+++ b/fs/ocfs2/ocfs2_ioctl.h
@@ -53,12 +53,23 @@ struct ocfs2_new_group_input {
 	__u32 frees;		/* Total free clusters in this group */
 	__u16 chain;		/* Chain for this group */
 	__u16 reserved1;
-	__u32 reserved2;
+	__u32 expand_clusters;
 };
 
-#define OCFS2_IOC_GROUP_EXTEND	_IOW('o', 1, int)
+struct ocfs2_extend_group_input {
+	__u32 new_clusters;
+	__u32 expand_clusters;
+};
+
+struct ocfs2_expand_volume_para {
+	struct ocfs2_extend_group_input *ei;
+	struct buffer_head *main_bm_bh;
+};
+
+#define OCFS2_IOC_GROUP_EXTEND	_IOW('o', 1, struct ocfs2_extend_group_input)
 #define OCFS2_IOC_GROUP_ADD	_IOW('o', 2,struct ocfs2_new_group_input)
 #define OCFS2_IOC_GROUP_ADD64	_IOW('o', 3,struct ocfs2_new_group_input)
+#define OCFS2_IOC_UPDATE_STOR_ATTR	_IOW('o', 5, unsigned int)
 
 /* Used to pass 2 file names to reflink. */
 struct reflink_arguments {
@@ -134,6 +145,13 @@ struct ocfs2_info_journal_size {
 	__u64 ij_journal_size;
 };
 
+struct ocfs2_info_journal_feature {
+	struct ocfs2_info_request ij_req;
+	__u32 opt_compat;
+	__u32 opt_incompat;
+	__u32 opt_ro_compat;
+};
+
 struct ocfs2_info_freeinode {
 	struct ocfs2_info_request ifi_req;
 	struct ocfs2_info_local_freeinode {
@@ -175,6 +193,7 @@ enum ocfs2_info_type {
 	OCFS2_INFO_UUID,
 	OCFS2_INFO_FS_FEATURES,
 	OCFS2_INFO_JOURNAL_SIZE,
+	OCFS2_INFO_JOURNAL_FEATURE,
 	OCFS2_INFO_FREEINODE,
 	OCFS2_INFO_FREEFRAG,
 	OCFS2_INFO_NUM_TYPES
@@ -230,5 +249,58 @@ struct ocfs2_move_extents {
 							 */
 
 #define OCFS2_IOC_MOVE_EXT	_IOW('o', 6, struct ocfs2_move_extents)
+#define OCFS2_IOC_RESORT_GD	_IOW('o', 7, void *)
+
+struct ocfs2_zero_range {
+	__u64 start;
+	__u64 len;
+	__u64 reserved[2];
+};
+
+#define OCFS2_IOC_ZERO		_IOW('z', 10, struct ocfs2_zero_range)
+#define OCFS2_ZERO_UPDATE	_IOW('z', 11, struct ocfs2_zero_range)
+
+#define OCFS2_DISK_NAME_LEN 32
+#define OCFS2_XCOPY_EXTENT_UNWRITTEN 0x01
+#define OCFS2_XCOPY_EXTENT_HOLE 0x02
+struct ocfs2_device_extent_rec {
+	__u8 disk_name[OCFS2_DISK_NAME_LEN];
+	__u64 pos;	/* file offset in bytes for the start */
+	__u64 phy_offset; /* physical offset in bytes for the start
+				* of the extent from the beginning of the disk */
+	__u64 length;   /* length in bytes for this extent */
+	__u8 flag; /*extent type(0: written, 1: unwritten, 2: hole)*/
+	__u64 dvd;
+};
+
+struct ocfs2_device_extent_list {
+	__u64 start;
+	__u64 len;
+	__u32 mapped_extents; /* number of extents that were mapped (out) */
+	__u32 max_extent_count;   /* size of extent array (in) */
+	struct ocfs2_device_extent_rec extent[0];
+};
+
+struct ocfs2_device_extent_info {
+	__u32 ei_mapped_extents;
+	__u32 ei_extent_count;
+	/* Start of ocfs2_device_extent_rec array */
+	struct ocfs2_device_extent_rec __user *ei_extents_start;
+};
+
+#define OCFS2_IOC_DEVADDR_QUERY	_IOW('x', 21, struct ocfs2_device_extent_list)
+#define OCFS2_XCOPY_MAX_EXTENTS	(UINT_MAX / sizeof(struct ocfs2_device_extent_rec))
+
+#define OCFS2_IOC_XCOPY_WRITE		_IOW('x', 22, struct ocfs2_device_extent_list)
+#define OCFS2_IOC_XCOPY_READ		_IOW('x', 23, struct ocfs2_device_extent_list)
+#define OCFS2_IOC_GET_XCOPY_FLAGS	_IOR('x', 24, unsigned int)
+#define OCFS2_IOC_GET_WS_FLAGS		_IOR('x', 25, unsigned int)
+
+#define OCFS2_IOC_SET_HIGH_PERF_DIO _IOW('a', 31, int)
+
+#define OCFS2_IOC_FENCE_SCSI _IOW('f', 41, void *)
+#define OCFS2_IOC_CLEAR_FENCE_SCSI _IOW('f', 42, void *)
+
+#define OCFS2_IOC_GET_FS_STATUS _IOR('g', 51, unsigned int)
 
 #endif /* OCFS2_IOCTL_H */
diff --git a/fs/ocfs2/ocfs2_trace.h b/fs/ocfs2/ocfs2_trace.h
index dc4bce1649c1..417489d6534b 100644
--- a/fs/ocfs2/ocfs2_trace.h
+++ b/fs/ocfs2/ocfs2_trace.h
@@ -1451,22 +1451,25 @@ DEFINE_OCFS2_ULL_ULL_ULL_EVENT(ocfs2_remove_inode_range);
 
 TRACE_EVENT(ocfs2_prepare_inode_for_write,
 	TP_PROTO(unsigned long long ino, unsigned long long saved_pos,
-		 unsigned long count, int wait),
-	TP_ARGS(ino, saved_pos, count, wait),
+		 int appending, unsigned long count, int *direct_io),
+	TP_ARGS(ino, saved_pos, appending, count, direct_io),
 	TP_STRUCT__entry(
 		__field(unsigned long long, ino)
 		__field(unsigned long long, saved_pos)
+		__field(int, appending)
 		__field(unsigned long, count)
-		__field(int, wait)
+		__field(int, direct_io)
 	),
 	TP_fast_assign(
 		__entry->ino = ino;
 		__entry->saved_pos = saved_pos;
+		__entry->appending = appending;
 		__entry->count = count;
-		__entry->wait = wait;
+		__entry->direct_io = direct_io ? *direct_io : -1;
 	),
-	TP_printk("%llu %llu %lu %d", __entry->ino,
-		  __entry->saved_pos, __entry->count, __entry->wait)
+	TP_printk("%llu %llu %d %lu %d", __entry->ino,
+		  __entry->saved_pos, __entry->appending, __entry->count,
+		  __entry->direct_io)
 );
 
 DEFINE_OCFS2_INT_EVENT(generic_file_read_iter_ret);
diff --git a/fs/ocfs2/resize.c b/fs/ocfs2/resize.c
index 24eb52f9059c..bc0e11284fed 100644
--- a/fs/ocfs2/resize.c
+++ b/fs/ocfs2/resize.c
@@ -12,19 +12,23 @@
 
 #include <linux/fs.h>
 #include <linux/types.h>
+#include <linux/delay.h>
+#include <linux/jiffies.h>
+#include <linux/sort.h>
 
 #include <cluster/masklog.h>
 
 #include "ocfs2.h"
 
 #include "alloc.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "inode.h"
 #include "journal.h"
 #include "super.h"
 #include "sysfile.h"
 #include "uptodate.h"
 #include "ocfs2_trace.h"
+#include "file.h"
 
 #include "buffer_head_io.h"
 #include "suballoc.h"
@@ -63,6 +67,8 @@ static u16 ocfs2_calc_new_backup_super(struct inode *inode,
 		lgd_cluster += old_bg_clusters;
 		if (lgd_cluster >= cluster)
 			continue;
+		if ((cluster % cl_cpg) > (gd->bg_bits - 1))
+			break;
 
 		if (set)
 			ocfs2_set_bit(cluster % cl_cpg,
@@ -81,7 +87,8 @@ static int ocfs2_update_last_group_and_inode(handle_t *handle,
 					     struct buffer_head *bm_bh,
 					     struct buffer_head *group_bh,
 					     u32 first_new_cluster,
-					     int new_clusters)
+					     int new_clusters,
+					     int expand_clusters)
 {
 	int ret = 0;
 	struct ocfs2_super *osb = OCFS2_SB(bm_inode->i_sb);
@@ -140,6 +147,7 @@ static int ocfs2_update_last_group_and_inode(handle_t *handle,
 	le32_add_cpu(&cr->c_free, num_bits);
 	le32_add_cpu(&fe->id1.bitmap1.i_total, num_bits);
 	le32_add_cpu(&fe->i_clusters, new_clusters);
+	fe->i_ex_clusters = cpu_to_le32(expand_clusters);
 
 	if (backups) {
 		le32_add_cpu(&cr->c_free, -1 * backups);
@@ -169,7 +177,7 @@ static int ocfs2_update_last_group_and_inode(handle_t *handle,
 	return ret;
 }
 
-static int update_backups(struct inode * inode, u32 clusters, char *data)
+int update_backups(struct inode *inode, u32 clusters, char *data)
 {
 	int i, ret = 0;
 	u32 cluster;
@@ -208,47 +216,169 @@ static int update_backups(struct inode * inode, u32 clusters, char *data)
 	return ret;
 }
 
-static void ocfs2_update_super_and_backups(struct inode *inode,
-					   int new_clusters)
+int ocfs2_update_super_and_backups(struct inode *inode,
+		unsigned int cmd, void *para)
 {
-	int ret;
+	int ret = 0;
 	u32 clusters = 0;
-	struct buffer_head *super_bh = NULL;
 	struct ocfs2_dinode *super_di = NULL;
+	struct ocfs2_dinode *bm_di = NULL;
 	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
+	struct ocfs2_expand_volume_para *ep;
+	struct ocfs2_extend_group_input *input;
+	unsigned int *flags;
+	u16 storage_attribute = 0, tmp_attr = 0;
 
 	/*
 	 * update the superblock last.
 	 * It doesn't matter if the write failed.
 	 */
-	ret = ocfs2_read_blocks_sync(osb, OCFS2_SUPER_BLOCK_BLKNO, 1,
-				     &super_bh);
+	mutex_lock(&osb->system_file_mutex);
+	ret = ocfs2_super_block_lock(osb, 1);
 	if (ret < 0) {
+		mutex_unlock(&osb->system_file_mutex);
 		mlog_errno(ret);
+		return ret;
+	}
+
+	super_di = (struct ocfs2_dinode *)osb->osb_super_bh->b_data;
+
+	switch (cmd) {
+	case OCFS2_ZERO_UPDATE:
+		if (le16_to_cpu(super_di->id2.i_super.s_storage_attribute) &
+				OCFS2_VAAI_WS_SUPPORT) {
+			mlog(ML_NOTICE,
+				"writesame performance has already existed on UUID %s.\n",
+				osb->uuid_str);
+			goto out;
+		}
+		super_di->id2.i_super.s_storage_attribute |= cpu_to_le16(OCFS2_VAAI_WS_SUPPORT);
+		spin_lock(&osb->osb_lock);
+		osb->osb_ws_support = 1;
+		spin_unlock(&osb->osb_lock);
+		break;
+	case OCFS2_IOC_UPDATE_STOR_ATTR:
+		if (para) {
+			storage_attribute = *(unsigned int *)para;
+			storage_attribute &= (OCFS2_VAAI_WS_SUPPORT | OCFS2_VAAI_XCOPY_SUPPORT |
+					OCFS2_FENCE_SCSI_SUPPORT);
+		}
+		tmp_attr = le16_to_cpu(super_di->id2.i_super.s_storage_attribute);
+		tmp_attr &= (OCFS2_VAAI_WS_SUPPORT | OCFS2_VAAI_XCOPY_SUPPORT |
+				OCFS2_FENCE_SCSI_SUPPORT);
+
+		if (!(storage_attribute & OCFS2_VAAI_WS_SUPPORT)) {
+			super_di->id2.i_super.s_storage_attribute &=
+					cpu_to_le16(~(OCFS2_VAAI_WS_SUPPORT |
+					OCFS2_VAAI_WS_SWITCH));
+			spin_lock(&osb->osb_lock);
+			osb->osb_ws_support = 0;
+			spin_unlock(&osb->osb_lock);
+		}
+
+		if (!(storage_attribute & OCFS2_VAAI_XCOPY_SUPPORT)) {
+			super_di->id2.i_super.s_storage_attribute &=
+					cpu_to_le16(~(OCFS2_VAAI_XCOPY_SUPPORT |
+					OCFS2_VAAI_XCOPY_SWITCH));
+			spin_lock(&osb->osb_lock);
+			osb->osb_xcopy_support = 0;
+			spin_unlock(&osb->osb_lock);
+		}
+
+		if (!(storage_attribute & OCFS2_FENCE_SCSI_SUPPORT)) {
+			super_di->id2.i_super.s_storage_attribute &=
+					cpu_to_le16(~(OCFS2_FENCE_SCSI_SUPPORT));
+			spin_lock(&osb->osb_lock);
+			osb->osb_fence_scsi_support = 0;
+			spin_unlock(&osb->osb_lock);
+		}
+
+		/* if 'storage_attribute' is equal to the value on disk,
+		 * no need set it to disk again.
+		 */
+		if (storage_attribute == tmp_attr) {
+			mlog(ML_NOTICE, "%s: only need set storage_attribute 0x%x in mem\n",
+				osb->uuid_str, storage_attribute);
+			goto out;
+		}
+		mlog(ML_NOTICE, "%s: set storage_attribute 0x%x on disk\n",
+				osb->uuid_str, storage_attribute);
+		break;
+	case OCFS2_IOC_GROUP_EXTEND:
+	case OCFS2_IOC_GROUP_ADD:
+	case OCFS2_IOC_GROUP_ADD64:
+		ep = para;
+		input = ep->ei;
+		bm_di = (struct ocfs2_dinode *)ep->main_bm_bh->b_data;
+		super_di->i_clusters = bm_di->i_clusters;
+		if (input)
+			super_di->i_ex_clusters = cpu_to_le32(input->expand_clusters);
+		break;
+	case OCFS2_IOC_FENCE_SCSI:
+		spin_lock(&osb->osb_lock);
+		osb->osb_fence_scsi_support = 1;
+		spin_unlock(&osb->osb_lock);
+
+		if (le16_to_cpu(super_di->id2.i_super.s_storage_attribute) &
+				OCFS2_FENCE_SCSI_SUPPORT)
+			goto out;
+		super_di->id2.i_super.s_storage_attribute |= cpu_to_le16(OCFS2_FENCE_SCSI_SUPPORT);
+		break;
+	case OCFS2_IOC_CLEAR_FENCE_SCSI:
+		spin_lock(&osb->osb_lock);
+		osb->osb_fence_scsi_support = 0;
+		spin_unlock(&osb->osb_lock);
+
+		if (!(le16_to_cpu(super_di->id2.i_super.s_storage_attribute) &
+				OCFS2_FENCE_SCSI_SUPPORT))
+			goto out;
+		super_di->id2.i_super.s_storage_attribute &= cpu_to_le16(~OCFS2_FENCE_SCSI_SUPPORT);
+		break;
+	case OCFS2_IOC_SET_HIGH_PERF_DIO:
+		flags = para;
+		spin_lock(&osb->osb_lock);
+		if (*flags && !ocfs2_supports_high_perf_dio(osb)) {
+			osb->s_feature_incompat |= OCFS2_FEATURE_INCOMPAT_HIGH_PERF_DIO;
+			super_di->id2.i_super.s_feature_incompat |=
+					cpu_to_le32(OCFS2_FEATURE_INCOMPAT_HIGH_PERF_DIO);
+		} else if (!(*flags) && ocfs2_supports_high_perf_dio(osb)) {
+			osb->s_feature_incompat &= ~OCFS2_FEATURE_INCOMPAT_HIGH_PERF_DIO;
+			super_di->id2.i_super.s_feature_incompat &=
+					cpu_to_le32(~OCFS2_FEATURE_INCOMPAT_HIGH_PERF_DIO);
+		}
+		spin_unlock(&osb->osb_lock);
+		break;
+	default:
+		mlog(ML_ERROR, "%s: invalid command %d\n", osb->uuid_str, cmd);
+		ret = -EINVAL;
 		goto out;
 	}
 
-	super_di = (struct ocfs2_dinode *)super_bh->b_data;
-	le32_add_cpu(&super_di->i_clusters, new_clusters);
 	clusters = le32_to_cpu(super_di->i_clusters);
 
-	ret = ocfs2_write_super_or_backup(osb, super_bh);
+	ret = ocfs2_write_super_or_backup(osb, osb->osb_super_bh);
 	if (ret < 0) {
 		mlog_errno(ret);
 		goto out;
 	}
 
-	if (OCFS2_HAS_COMPAT_FEATURE(osb->sb, OCFS2_FEATURE_COMPAT_BACKUP_SB))
-		ret = update_backups(inode, clusters, super_bh->b_data);
+	if (OCFS2_HAS_COMPAT_FEATURE(osb->sb, OCFS2_FEATURE_COMPAT_BACKUP_SB)) {
+		ret = update_backups(inode, clusters, osb->osb_super_bh->b_data);
+		if (ret < 0) {
+			mlog_errno(ret);
+			goto out;
+		}
+	}
 
 out:
-	brelse(super_bh);
 	if (ret)
 		printk(KERN_WARNING "ocfs2: Failed to update super blocks on %s"
-			" during fs resize. This condition is not fatal,"
-			" but fsck.ocfs2 should be run to fix it\n",
-			osb->dev_str);
-	return;
+			"(uuid %s) during fs resize. This condition is not fatal,"
+			" but fsck.ocfs2 should be run to fix it or resize the volume again\n",
+			osb->dev_str, osb->uuid_str);
+	ocfs2_super_block_unlock(osb, 1);
+	mutex_unlock(&osb->system_file_mutex);
+	return ret;
 }
 
 /*
@@ -256,10 +386,13 @@ static void ocfs2_update_super_and_backups(struct inode *inode,
  * point is only used to extend the current filesystem to the end of the last
  * existing group.
  */
-int ocfs2_group_extend(struct inode * inode, int new_clusters)
+int ocfs2_group_extend(struct inode *inode,
+		unsigned int cmd,
+		struct ocfs2_extend_group_input *input)
 {
-	int ret;
+	int ret, update_sb = 0;
 	handle_t *handle;
+	journal_t *journal;
 	struct buffer_head *main_bm_bh = NULL;
 	struct buffer_head *group_bh = NULL;
 	struct inode *main_bm_inode = NULL;
@@ -269,13 +402,12 @@ int ocfs2_group_extend(struct inode * inode, int new_clusters)
 	u16 cl_bpc;
 	u32 first_new_cluster;
 	u64 lgd_blkno;
+	struct ocfs2_expand_volume_para ep = {0};
 
 	if (ocfs2_is_hard_readonly(osb) || ocfs2_is_soft_readonly(osb))
 		return -EROFS;
 
-	if (new_clusters < 0)
-		return -EINVAL;
-	else if (new_clusters == 0)
+	if (input->new_clusters == 0)
 		return 0;
 
 	main_bm_inode = ocfs2_get_system_file_inode(osb,
@@ -298,14 +430,22 @@ int ocfs2_group_extend(struct inode * inode, int new_clusters)
 	fe = (struct ocfs2_dinode *)main_bm_bh->b_data;
 
 	/* main_bm_bh is validated by inode read inside ocfs2_inode_lock(),
-	 * so any corruption is a code bug. */
-	BUG_ON(!OCFS2_IS_VALID_DINODE(fe));
+	 * but cache may be destroyed when using dd command, so main_bm_bh
+	 * may be corrupted then */
+	if (!OCFS2_IS_VALID_DINODE(fe)) {
+		ocfs2_error(osb->sb, "Invalid dinode #%llu: signature = %.*s",
+				(unsigned long long)OCFS2_I(main_bm_inode)->ip_blkno, 7,
+				fe->i_signature);
+		ret = -EROFS;
+		goto out_unlock;
+	}
 
 	if (le16_to_cpu(fe->id2.i_chain.cl_cpg) !=
 		ocfs2_group_bitmap_size(osb->sb, 0,
 					osb->s_feature_incompat) * 8) {
-		mlog(ML_ERROR, "The disk is too old and small. "
-		     "Force to do offline resize.");
+		mlog(ML_ERROR,
+			"%s: The disk is too old and small. Force to do offline resize.\n",
+			osb->uuid_str);
 		ret = -EINVAL;
 		goto out_unlock;
 	}
@@ -323,15 +463,15 @@ int ocfs2_group_extend(struct inode * inode, int new_clusters)
 	group = (struct ocfs2_group_desc *)group_bh->b_data;
 
 	cl_bpc = le16_to_cpu(fe->id2.i_chain.cl_bpc);
-	if (le16_to_cpu(group->bg_bits) / cl_bpc + new_clusters >
+	if (le16_to_cpu(group->bg_bits) / cl_bpc + input->new_clusters >
 		le16_to_cpu(fe->id2.i_chain.cl_cpg)) {
+		mlog(ML_ERROR, "%s: invalid new_clusters.\n", osb->uuid_str);
 		ret = -EINVAL;
 		goto out_unlock;
 	}
 
-
 	trace_ocfs2_group_extend(
-	     (unsigned long long)le64_to_cpu(group->bg_blkno), new_clusters);
+	     (unsigned long long)le64_to_cpu(group->bg_blkno), input->new_clusters);
 
 	handle = ocfs2_start_trans(osb, OCFS2_GROUP_EXTEND_CREDITS);
 	if (IS_ERR(handle)) {
@@ -344,16 +484,29 @@ int ocfs2_group_extend(struct inode * inode, int new_clusters)
 	ret = ocfs2_update_last_group_and_inode(handle, main_bm_inode,
 						main_bm_bh, group_bh,
 						first_new_cluster,
-						new_clusters);
+						input->new_clusters,
+						input->expand_clusters);
 	if (ret) {
 		mlog_errno(ret);
 		goto out_commit;
 	}
 
-	ocfs2_update_super_and_backups(main_bm_inode, new_clusters);
+	ep.ei = input;
+	ep.main_bm_bh = main_bm_bh;
+	update_sb = 1;
 
 out_commit:
 	ocfs2_commit_trans(osb, handle);
+	if (!ret) {
+		journal = osb->journal->j_journal;
+		ret = jbd2_journal_force_commit(journal);
+		if (ret < 0) {
+			mlog_errno(ret);
+			goto out_unlock;
+		}
+	}
+	if (update_sb)
+		ocfs2_update_super_and_backups(main_bm_inode, cmd, (void *)&ep);
 out_unlock:
 	brelse(group_bh);
 	brelse(main_bm_bh);
@@ -377,6 +530,7 @@ static int ocfs2_check_new_group(struct inode *inode,
 	struct ocfs2_group_desc *gd =
 		(struct ocfs2_group_desc *)group_bh->b_data;
 	u16 cl_bpc = le16_to_cpu(di->id2.i_chain.cl_bpc);
+	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
 
 	ret = ocfs2_check_group_descriptor(inode->i_sb, di, group_bh);
 	if (ret)
@@ -384,18 +538,24 @@ static int ocfs2_check_new_group(struct inode *inode,
 
 	ret = -EINVAL;
 	if (le16_to_cpu(gd->bg_chain) != input->chain)
-		mlog(ML_ERROR, "Group descriptor # %llu has bad chain %u "
+		mlog(ML_ERROR,
+		     "%s: Group descriptor # %llu has bad chain %u "
 		     "while input has %u set.\n",
+		     osb->uuid_str,
 		     (unsigned long long)le64_to_cpu(gd->bg_blkno),
 		     le16_to_cpu(gd->bg_chain), input->chain);
 	else if (le16_to_cpu(gd->bg_bits) != input->clusters * cl_bpc)
-		mlog(ML_ERROR, "Group descriptor # %llu has bit count %u but "
+		mlog(ML_ERROR,
+		     "%s: Group descriptor # %llu has bit count %u but "
 		     "input has %u clusters set\n",
+		     osb->uuid_str,
 		     (unsigned long long)le64_to_cpu(gd->bg_blkno),
 		     le16_to_cpu(gd->bg_bits), input->clusters);
 	else if (le16_to_cpu(gd->bg_free_bits_count) != input->frees * cl_bpc)
-		mlog(ML_ERROR, "Group descriptor # %llu has free bit count %u "
+		mlog(ML_ERROR,
+		     "%s: Group descriptor # %llu has free bit count %u "
 		     "but it should have %u set\n",
+		     osb->uuid_str,
 		     (unsigned long long)le64_to_cpu(gd->bg_blkno),
 		     le16_to_cpu(gd->bg_bits),
 		     input->frees * cl_bpc);
@@ -417,27 +577,31 @@ static int ocfs2_verify_group_and_input(struct inode *inode,
 	u32 cluster = ocfs2_blocks_to_clusters(inode->i_sb, input->group);
 	u32 total_clusters = le32_to_cpu(di->i_clusters);
 	int ret = -EINVAL;
+	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
 
 	if (cluster < total_clusters)
-		mlog(ML_ERROR, "add a group which is in the current volume.\n");
+		mlog(ML_ERROR, "%s: add a group which is in the current volume.\n", osb->uuid_str);
 	else if (input->chain >= cl_count)
-		mlog(ML_ERROR, "input chain exceeds the limit.\n");
+		mlog(ML_ERROR, "%s: input chain exceeds the limit.\n", osb->uuid_str);
 	else if (next_free != cl_count && next_free != input->chain)
 		mlog(ML_ERROR,
-		     "the add group should be in chain %u\n", next_free);
+		     "%s: the add group should be in chain %u\n", osb->uuid_str, next_free);
 	else if (total_clusters + input->clusters < total_clusters)
-		mlog(ML_ERROR, "add group's clusters overflow.\n");
+		mlog(ML_ERROR, "%s: add group's clusters overflow.\n", osb->uuid_str);
 	else if (input->clusters > cl_cpg)
-		mlog(ML_ERROR, "the cluster exceeds the maximum of a group\n");
+		mlog(ML_ERROR, "%s: the cluster exceeds the maximum of a group\n", osb->uuid_str);
 	else if (input->frees > input->clusters)
-		mlog(ML_ERROR, "the free cluster exceeds the total clusters\n");
+		mlog(ML_ERROR, "%s: the free cluster exceeds the total clusters\n", osb->uuid_str);
 	else if (total_clusters % cl_cpg != 0)
 		mlog(ML_ERROR,
-		     "the last group isn't full. Use group extend first.\n");
+		     "%s: the last group isn't full. Use group extend first.\n", osb->uuid_str);
 	else if (input->group != ocfs2_which_cluster_group(inode, cluster))
-		mlog(ML_ERROR, "group blkno is invalid\n");
+		mlog(ML_ERROR, "%s: group blkno is invalid\n", osb->uuid_str);
 	else if ((ret = ocfs2_check_new_group(inode, di, input, group_bh)))
-		mlog(ML_ERROR, "group descriptor check failed.\n");
+		mlog(ML_ERROR, "%s: group descriptor check failed.\n", osb->uuid_str);
+	else if (input->expand_clusters < total_clusters)
+		mlog(ML_ERROR, "%s: invalid expand_clusters(%u) and total_clusters(%u).\n",
+				osb->uuid_str, input->expand_clusters, total_clusters);
 	else
 		ret = 0;
 
@@ -445,10 +609,11 @@ static int ocfs2_verify_group_and_input(struct inode *inode,
 }
 
 /* Add a new group descriptor to global_bitmap. */
-int ocfs2_group_add(struct inode *inode, struct ocfs2_new_group_input *input)
+int ocfs2_group_add(struct inode *inode, unsigned int cmd, struct ocfs2_new_group_input *input)
 {
-	int ret;
+	int ret, update_sb = 0;
 	handle_t *handle;
+	journal_t *journal;
 	struct buffer_head *main_bm_bh = NULL;
 	struct inode *main_bm_inode = NULL;
 	struct ocfs2_dinode *fe = NULL;
@@ -459,6 +624,8 @@ int ocfs2_group_add(struct inode *inode, struct ocfs2_new_group_input *input)
 	struct ocfs2_chain_rec *cr;
 	u16 cl_bpc;
 	u64 bg_ptr;
+	struct ocfs2_expand_volume_para ep = {0};
+	struct ocfs2_extend_group_input support_input;
 
 	if (ocfs2_is_hard_readonly(osb) || ocfs2_is_soft_readonly(osb))
 		return -EROFS;
@@ -485,21 +652,21 @@ int ocfs2_group_add(struct inode *inode, struct ocfs2_new_group_input *input)
 	if (le16_to_cpu(fe->id2.i_chain.cl_cpg) !=
 		ocfs2_group_bitmap_size(osb->sb, 0,
 					osb->s_feature_incompat) * 8) {
-		mlog(ML_ERROR, "The disk is too old and small."
-		     " Force to do offline resize.");
+		mlog(ML_ERROR,
+			"%s: The disk is too old and small. Force to do offline resize.\n",
+			osb->uuid_str);
 		ret = -EINVAL;
 		goto out_unlock;
 	}
 
 	ret = ocfs2_read_blocks_sync(osb, input->group, 1, &group_bh);
 	if (ret < 0) {
-		mlog(ML_ERROR, "Can't read the group descriptor # %llu "
-		     "from the device.", (unsigned long long)input->group);
+		mlog(ML_ERROR,
+			"%s: Can't read the group descriptor # %llu from the device.\n",
+			osb->uuid_str, (unsigned long long)input->group);
 		goto out_unlock;
 	}
 
-	ocfs2_set_new_buffer_uptodate(INODE_CACHE(inode), group_bh);
-
 	ret = ocfs2_verify_group_and_input(main_bm_inode, fe, input, group_bh);
 	if (ret) {
 		mlog_errno(ret);
@@ -553,6 +720,7 @@ int ocfs2_group_add(struct inode *inode, struct ocfs2_new_group_input *input)
 	le32_add_cpu(&fe->id1.bitmap1.i_used,
 		     (input->clusters - input->frees) * cl_bpc);
 	le32_add_cpu(&fe->i_clusters, input->clusters);
+	fe->i_ex_clusters = cpu_to_le32(input->expand_clusters);
 
 	ocfs2_journal_dirty(handle, main_bm_bh);
 
@@ -562,10 +730,24 @@ int ocfs2_group_add(struct inode *inode, struct ocfs2_new_group_input *input)
 	spin_unlock(&OCFS2_I(main_bm_inode)->ip_lock);
 	i_size_write(main_bm_inode, le64_to_cpu(fe->i_size));
 
-	ocfs2_update_super_and_backups(main_bm_inode, input->clusters);
+	support_input.new_clusters = input->clusters;
+	support_input.expand_clusters = input->expand_clusters;
+	ep.ei = &support_input;
+	ep.main_bm_bh = main_bm_bh;
+	update_sb = 1;
 
 out_commit:
 	ocfs2_commit_trans(osb, handle);
+	if (!ret) {
+		journal = osb->journal->j_journal;
+		ret = jbd2_journal_force_commit(journal);
+		if (ret < 0) {
+			mlog_errno(ret);
+			goto out_unlock;
+		}
+	}
+	if (update_sb)
+		ocfs2_update_super_and_backups(main_bm_inode, cmd, (void *)&ep);
 
 out_free_group_bh:
 	brelse(group_bh);
@@ -582,3 +764,190 @@ int ocfs2_group_add(struct inode *inode, struct ocfs2_new_group_input *input)
 out:
 	return ret;
 }
+
+static int cmp_func(const void *a, const void *b)
+{
+	int a_free_bit, b_free_bit;
+	struct ocfs2_group_desc *gd;
+
+	gd = (struct ocfs2_group_desc *)((*((struct buffer_head **)a))->b_data);
+	a_free_bit = le16_to_cpu(gd->bg_free_bits_count);
+
+	gd = (struct ocfs2_group_desc *)((*((struct buffer_head **)b))->b_data);
+	b_free_bit = le16_to_cpu(gd->bg_free_bits_count);
+
+	return b_free_bit - a_free_bit;
+}
+
+static int ocfs2_resort_chain(struct ocfs2_super *osb,
+			      struct inode *bm_inode,
+			      struct buffer_head *bm_bh,
+			      int num)
+{
+	int ret = 0, i;
+	u32 num_gds, max_chain_len, chain_len = 0;
+	u16 cpg;
+	u64 blkno;
+	journal_t *journal;
+	handle_t *handle = NULL;
+	struct ocfs2_chain_list *cl = NULL;
+	struct ocfs2_dinode *fe = NULL;
+	struct buffer_head **bh = NULL;
+	struct ocfs2_group_desc *gd_next = NULL;
+	struct ocfs2_group_desc *gd = NULL;
+
+	fe = (struct ocfs2_dinode *)bm_bh->b_data;
+	cl = &(fe->id2.i_chain);
+	cpg = le16_to_cpu(cl->cl_cpg);
+	num_gds = (le32_to_cpu(fe->i_clusters) + cpg - 1) / cpg;
+	max_chain_len = (num_gds + le16_to_cpu(cl->cl_count) - 1) / le16_to_cpu(cl->cl_count);
+	blkno = le64_to_cpu(cl->cl_recs[num].c_blkno);
+
+	/* nothing to do when there's only one gd in each chain */
+	if (max_chain_len < 2)
+		return 0;
+
+	bh = kcalloc(max_chain_len, sizeof(struct buffer_head *), GFP_NOFS);
+	if (!bh) {
+		ret = -ENOMEM;
+		mlog_errno(ret);
+		return ret;
+	}
+
+	while (chain_len < max_chain_len && blkno) {
+		ret = ocfs2_read_group_descriptor(bm_inode, fe, blkno, &bh[chain_len]);
+		if (ret < 0) {
+			mlog(ML_ERROR,
+				"%s: Can't read the group descriptor # %llu from the device.\n",
+				osb->uuid_str, (unsigned long long)blkno);
+			goto out_free;
+		}
+		gd = (struct ocfs2_group_desc *)bh[chain_len]->b_data;
+		blkno = le64_to_cpu(gd->bg_next_group);
+		chain_len++;
+	}
+
+	/* may be still only one gd in some chains, minimize gd write as much as possible */
+	if (chain_len < 2)
+		goto out_free;
+
+	sort(bh, chain_len, sizeof(struct buffer_head *), &cmp_func, NULL);
+
+	/* global_bitmap dinode and chain_len gds update */
+	handle = ocfs2_start_trans(osb, chain_len + 1);
+	if (IS_ERR(handle)) {
+		mlog_errno(PTR_ERR(handle));
+		ret = -EINVAL;
+		goto out_free;
+	}
+
+	ret = ocfs2_journal_access_di(handle, INODE_CACHE(bm_inode),
+			bm_bh, OCFS2_JOURNAL_ACCESS_WRITE);
+	if (ret < 0) {
+		mlog_errno(ret);
+		goto out_commit;
+	}
+
+	for (i = 0; i < chain_len; i++) {
+		ret = ocfs2_journal_access_gd(handle, INODE_CACHE(bm_inode),
+				bh[i], OCFS2_JOURNAL_ACCESS_WRITE);
+		if (ret < 0) {
+			mlog_errno(ret);
+			goto out_commit;
+		}
+	}
+
+	gd = (struct ocfs2_group_desc *)bh[0]->b_data;
+	cl->cl_recs[num].c_blkno = gd->bg_blkno;
+	ocfs2_journal_dirty(handle, bm_bh);
+
+	for (i = 1; i < chain_len; i++) {
+		gd_next = (struct ocfs2_group_desc *)bh[i]->b_data;
+		gd->bg_next_group = gd_next->bg_blkno;
+		ocfs2_journal_dirty(handle, bh[i - 1]);
+		gd = gd_next;
+	}
+
+	gd->bg_next_group = 0;
+	ocfs2_journal_dirty(handle, bh[i - 1]);
+
+out_commit:
+	ocfs2_commit_trans(osb, handle);
+	if (!ret) {
+		journal = osb->journal->j_journal;
+		ret = jbd2_journal_force_commit(journal);
+		if (ret < 0)
+			mlog_errno(ret);
+	}
+
+out_free:
+	for (i = 0; i < chain_len; i++)
+		brelse(bh[i]);
+	kfree(bh);
+
+	return ret;
+}
+
+int ocfs2_resort_global_bitmap(struct inode *inode)
+{
+	int ret, cur_chain, total_chain;
+	struct ocfs2_dinode *fe = NULL;
+	struct ocfs2_chain_list *cl;
+	struct inode *global_bitmap_inode = NULL;
+	struct buffer_head *bh = NULL;
+	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
+
+	if (ocfs2_is_hard_readonly(osb) || ocfs2_is_soft_readonly(osb))
+		return -EROFS;
+
+	global_bitmap_inode = ocfs2_get_system_file_inode(osb,
+			GLOBAL_BITMAP_SYSTEM_INODE, OCFS2_INVALID_SLOT);
+	if (!global_bitmap_inode) {
+		mlog_errno(-EINVAL);
+		return -EINVAL;
+	}
+
+	cur_chain = total_chain = 0;
+	do {
+		inode_lock(global_bitmap_inode);
+		ret = ocfs2_inode_lock(global_bitmap_inode, &bh, 1);
+		if (ret < 0) {
+			mlog_errno(ret);
+			inode_unlock(global_bitmap_inode);
+			goto out;
+		}
+
+		fe = (struct ocfs2_dinode *)bh->b_data;
+		cl = &(fe->id2.i_chain);
+		total_chain = le16_to_cpu(cl->cl_next_free_rec);
+
+		if (cur_chain >= total_chain) {
+			ocfs2_inode_unlock(global_bitmap_inode, 1);
+			inode_unlock(global_bitmap_inode);
+			mlog(ML_ERROR,
+				"Chain %d exceeds the max chain length %d, group block (%llu)",
+				cur_chain, total_chain, (unsigned long long)bh->b_blocknr);
+			goto out;
+		}
+
+		ret = ocfs2_resort_chain(osb, global_bitmap_inode, bh, cur_chain);
+		if (ret) {
+			ocfs2_inode_unlock(global_bitmap_inode, 1);
+			inode_unlock(global_bitmap_inode);
+			mlog(ML_ERROR, "sort chain %d failed\n", cur_chain);
+			goto out;
+		}
+
+		cur_chain++;
+		ocfs2_inode_unlock(global_bitmap_inode, 1);
+		inode_unlock(global_bitmap_inode);
+		brelse(bh);
+		bh = NULL;
+	} while (cur_chain < total_chain);
+
+out:
+	if (bh)
+		brelse(bh);
+	iput(global_bitmap_inode);
+	return ret;
+}
diff --git a/fs/ocfs2/resize.h b/fs/ocfs2/resize.h
index 0af0c023042c..4dd092c14656 100644
--- a/fs/ocfs2/resize.h
+++ b/fs/ocfs2/resize.h
@@ -12,7 +12,14 @@
 #ifndef OCFS2_RESIZE_H
 #define OCFS2_RESIZE_H
 
-int ocfs2_group_extend(struct inode * inode, int new_clusters);
-int ocfs2_group_add(struct inode *inode, struct ocfs2_new_group_input *input);
+int ocfs2_group_extend(struct inode *inode, unsigned int cmd,
+		struct ocfs2_extend_group_input *input);
+int ocfs2_group_add(struct inode *inode, unsigned int cmd,
+		struct ocfs2_new_group_input *input);
+
+int ocfs2_update_super_and_backups(struct inode *inode,
+		unsigned int cmd, void *para);
+int update_backups(struct inode *inode, u32 clusters, char *data);
+int ocfs2_resort_global_bitmap(struct inode *inode);
 
 #endif /* OCFS2_RESIZE_H */
-- 
2.17.1

